File: dotnet/src/Skills/Skills.Core/FileIOSkill.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.ComponentModel;
using System.IO;
using System.Text;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.SkillDefinition;

namespace Microsoft.SemanticKernel.Skills.Core;

/// <summary>
/// Read and write from a file.
/// </summary>
/// <example>
/// Usage: kernel.ImportSkill("file", new FileIOSkill());
/// Examples:
/// {{file.readAsync $path }} => "hello world"
/// {{file.writeAsync}}
/// </example>
public sealed class FileIOSkill
{
    /// <summary>
    /// Read a file
    /// </summary>
    /// <example>
    /// {{file.readAsync $path }} => "hello world"
    /// </example>
    /// <param name="path"> Source file </param>
    /// <returns> File content </returns>
    [SKFunction, Description("Read a file")]
    public async Task<string> ReadAsync([Description("Source file")] string path)
    {
        using var reader = File.OpenText(path);
        return await reader.ReadToEndAsync().ConfigureAwait(false);
    }

    /// <summary>
    /// Write a file
    /// </summary>
    /// <example>
    /// {{file.writeAsync}}
    /// </example>
    /// <param name="path">The destination file path</param>
    /// <param name="content">The file content to write</param>
    /// <returns> An awaitable task </returns>
    [SKFunction, Description("Write a file")]
    public async Task WriteAsync(
        [Description("Destination file")] string path,
        [Description("File content")] string content)
    {
        byte[] text = Encoding.UTF8.GetBytes(content);
        if (File.Exists(path) && File.GetAttributes(path).HasFlag(FileAttributes.ReadOnly))
        {
            // Most environments will throw this with OpenWrite, but running inside docker on Linux will not.
            throw new UnauthorizedAccessException($"File is read-only: {path}");
        }

        using var writer = File.OpenWrite(path);
        await writer.WriteAsync(text, 0, text.Length).ConfigureAwait(false);
    }
}

File: java/plugins/semantickernel-plugin-core/src/main/java/com/microsoft/semantickernel/coreskills/FileIOSkill.java
// Copyright (c) Microsoft. All rights reserved.
package com.microsoft.semantickernel.coreskills;

import com.microsoft.semantickernel.skilldefinition.annotations.DefineSKFunction;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionInputAttribute;
import com.microsoft.semantickernel.skilldefinition.annotations.SKFunctionParameters;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

/**
 * Read and write from a file.
 *
 * <p>Usage: kernel.ImportSkill("file", new FileIOSkill());
 *
 * <p>Examples:
 *
 * <pre>{@code
 * file.readAsync($path) => "hello world"
 * file.writeAsync()
 * }</pre>
 */
public class FileIOSkill {

    /**
     * Read a file.
     *
     * <p>Example:
     *
     * <pre>{@code
     * file.readAsync($path) => "hello world"
     * }</pre>
     *
     * @param path Source file.
     * @return File content.
     */
    @DefineSKFunction(description = "Read a file", name = "readAsync")
    public String readFileAsync(
            @SKFunctionInputAttribute(description = "Source file") String path,
            @SKFunctionParameters(
                            name = "charset",
                            description = "Character set to use to read the file",
                            defaultValue = "UTF-8")
                    String charset)
            throws IOException {
        Path filePath = Paths.get(path);
        byte[] fileBytes;
        fileBytes = Files.readAllBytes(filePath);
        String fileContents = new String(fileBytes, Charset.forName(charset));
        return fileContents;
    }

    /**
     * Write a file.
     *
     * <p>Example:
     *
     * <pre>{@code
     * file.writeAsync()
     * }</pre>
     *
     * @param path The destination file.
     * @param content Contains the 'content' to write
     */
    @DefineSKFunction(description = "Write a file", name = "writeAsync")
    public void writeFileAsync(
            @SKFunctionInputAttribute(description = "Destination file") String path,
            @SKFunctionParameters(
                            name = "content",
                            description = "File content",
                            defaultValue = "",
                            type = String.class)
                    String content,
            @SKFunctionParameters(
                            name = "charset",
                            description = "Character set to use to read the file",
                            defaultValue = "UTF-8")
                    String charset)
            throws IOException {
        Path filePath = Paths.get(path);
        Files.write(filePath, content.getBytes(charset));
    }
}
